"dotnet test project.json" fails in a directory with a test project.json in it

eerhardt opened the pull request:
When the current directory contains a valid project.json, and the user just says `dotnet test project.json`, normalizing the path fails becuase we end up calling Path.GetFullPath with an empty string.

To fix this, if the directory of the file is empty, use the current directory.

Fix #2691

@brthor @troydai

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dotnet/cli/2700)
<!-- Reviewable:end -->

brthor commented on src/Microsoft.DotNet.ProjectModel/WorkspaceContext.cs:
```diff
@@ -445,7 +445,13 @@ private static string NormalizeProjectPath(string path)
                 string.Equals(Path.GetFileName(path), Project.FileName, StringComparison.OrdinalIgnoreCase))
             {
-                return Path.GetFullPath(Path.GetDirectoryName(path));
+                string directoryName = Path.GetDirectoryName(path);
```
should we add a dotnet test, test for this?

eerhardt replied:
I'm not sure we need to test everything end to end. This same bug exists for publish, and run. Should we add tests for all of them?

My thought was to unit test this at the common place, and all the callsites don't need to be tested.

brthor replied:
I do think that would be appropriate. 

These functional tests have been pretty valuable, and this is a callsite level issue which could be regressed without failing this unit tests.

eerhardt commented:
@brthor - it turns out adding those extra tests was beneficial.  It found another code path in "dotnet pack" that didn't go through the code path I fixed.

I added tests for all 5 commands that take a project path.  PTAL.

livarcocc commented on src/Microsoft.DotNet.ProjectModel/ProjectReader.cs:
```diff
@@ -246,6 +243,38 @@ public Project ReadProject(Stream stream, string projectName, string projectPath
             return project;
         }
 
+        internal static string NormalizeProjectDirectoryPath(string path)
```
ProjectReader do not feel like the right place for these methods for me.

eerhardt replied:
Can you recommend a better place?

livarcocc replied:
A ProjectPath internal class?

eerhardt replied:
Added a new ProjectPath class.

livarcocc commented on src/Microsoft.DotNet.ProjectModel/ProjectReader.cs:
```diff
@@ -246,6 +243,38 @@ public Project ReadProject(Stream stream, string projectName, string projectPath
+                return Path.GetFullPath(path);
+            }
+
+            return null;
```
would string.Empty be better?

eerhardt replied:
Not really.  The general rule I've used in the past is 'null' and 'empty string' should be handled the same way.  That's why there is a "IsNullOrEmpty" method. So one is just as good as the other.

Also - this is what the existing code did.

livarcocc replied:
Well.. null can crash the app while string.empty can't. Right? I mean, I don't see: that's what we have been doing as a real argument to let's continue doing it.

eerhardt replied:
> Well.. null can crash the app while string.empty can't. Right?

It's totally up to the caller to crash the app.  Just returning null doesn't crash the app.  The caller could also do `str[0]` which would crash the app.

> I mean, I don't see: that's what we have been doing as a real argument to let's continue doing it.

It is a real argument when you have callers that do this:

```C#
            var normalizedPath = ProjectReader.NormalizeProjectDirectoryPath(projectPath);
            if (normalizedPath == null)
            {
                return null;
            }
```

livarcocc replied:
No. IF you return null you put the onus on the caller to crash the app, while if you return something (like string.empty), you don't give him the chance. Well, make it harder.
There is a lot of literature out there defending not returning null makes for cleaner and better code. I can direct you to some if you want.

eerhardt replied:
OK, I'd like to read it.

But I'm still returning null here.  :smile:

livarcocc replied:
Why? Do you want to do the opposite and convince me that null is better? because it seems to me that return string.Empty; is a very small chance that should cost literally nothing do make.

eerhardt replied:
The fact that "return string.Empty" occurs exactly 3 times in our current code base tells me that this isn't a general convention.

1. and one of those times is in a method that isn't even called.
2. And the other 2 are in the same method

Another reason is performane-wise returning `null` is a faster check for the caller.  You don't have to dereference the string and check for Length == 0.

> because it seems to me that return string.Empty; is a very small chance that should cost literally nothing do make.

I've already shown you that there are callers of this method that explicitly check for null.  So `cost literally nothing` doesn't apply here.

livarcocc replied:
All right, I will take a stab at fixing it once your changes are in then.

livarcocc commented on src/Microsoft.DotNet.ProjectModel/ProjectReader.cs:
```diff
@@ -246,6 +243,38 @@ public Project ReadProject(Stream stream, string projectName, string projectPath
 
+        internal static string NormalizeProjectDirectoryPath(string path)
+        {
+            if (File.Exists(path) &&
```
It feels to me that the readability of this method could be improved a lot by simply extracting contents of the if conditions into nicely named methods. Likewise for the contents of the body of the ifs.

eerhardt replied:
Agreed the readability could be improved, but I don't feel this is the right time to be making that refactoring. If you feel strongly, please log an issue for it.

piotrpMSFT replied:
Is the issue "clean up code from PR 2700"? ;)

eerhardt replied:
This code was moved from WorkspaceContext to here with 1 line changing in it.

Normally at the end of a milestone you don't make unnecessary refactorings.  You usually have other, more important, things to attend to.  And you want to minimize churn in the code base, limit bugs.

livarcocc replied:
That's true when you don't have a set of tests and unit tests protecting you from breaking things. Plus, I suggested just a couple of extract methods refactoring to make this method more readable. Just use resharper, mark the code and say extract method. Give it a nice descriptive name and call it a day.

eerhardt replied:
Refactored and simplified the logic.

troydai commented on src/Microsoft.DotNet.ProjectModel/ProjectPath.cs:
```diff
@@ -0,0 +1,48 @@
+                return fullPath;
+            }
+
+            return null;
```
I was about to say: "returns `null`? how reckless" and then I found out we did that before. "well, OK then..."

troydai commented on src/Microsoft.DotNet.ProjectModel/ProjectPath.cs:
```diff
@@ -0,0 +1,48 @@
+
+namespace Microsoft.DotNet.ProjectModel
+{
+    internal static class ProjectPath
```
The name is a little bit to generic. I look at this and think of a structure represents a path. It acts more like a helper class. So how about `ProjectPathHelper`?

troydai commented:
Besides my nitpickings looks good to me.

troydai commented:
:shipit:

livarcocc commented:
I would say squash, at the least the address PR feedback commits and then :shipit:
